Feature 675 Continuous Integration Documentation (#1409)
dtcenter/METplus: docs/Contributors_Guide/continuous_integration.rst

**********************
Continuous Integration
**********************

METplus utilizes GitHub Actions to run processes automatically when changes
are pushed to GitHub. These tasks include:

* Building documentation to catch warnings/errors
* Building a Docker image to run tests
* Creating/Updating Docker data volumes with new input data used for tests
* Running unit tests
* Running use cases
* Comparing use case output to truth data
* Creating/Updating Docker data volumes with truth data to use in comparisons

GitHub Actions Workflows
========================

GitHub Actions runs workflows defined by files in the **.github/workflows**
directory of a GitHub repository.
Files with the .yml suffix are parsed and GitHub Actions will
trigger a workflow run if the triggering criteria is met.
It can run multiple jobs in parallel or serially depending on dependency rules
that can be set. Each job can run a series of commands or scripts called steps.

> Job steps can include "actions" with can be used to perform tasks. Many useful

Reviewer: ```suggestion
Job steps can include "actions" which can be used to perform tasks. Many useful
```

actions are provided by GitHub and external collaborators. Developers can also
write their own custom actions to perform complex tasks to simplify a workflow.

> 

Reviewer: Consider noting here that a single push can result in multiple workflows being triggered.

All workflow runs can be seen on the Actions tab of the repository. Each workflow run is identified by the branch for which it was invoked as well as the corresponding commit message on that branch. In general, a green check mark indicates that all checks for that workflow run passed. A red X indicates that at least one of the workflow checks failed.

Any workflow runs associated with a PR can also be seen on the "Checks" tab of that PR.

Consider adding 1 or 2 images to illustrate.

.. figure:: figure/ci-doc-error.png

A summary of the lines that contain WARNING or ERROR are output in the
GitHub Actions log for easy access.
The warnings.log file is also made available as a GitHub Actions
artifact so it can be downloaded and reviewed. Artifacts can be found
at the bottom of the workflow summary page when the workflow has completed.

.. figure:: figure/ci-doc-artifacts.png

Release Published (release_published.yml) - DEPRECATED
------------------------------------------------------

**This workflow is no longer be required, as Slack now has GitHub integration
to automatically create posts on certain events.** The workflow YAML file
is still found in the repository for reference, but the workflow has been
disabled via the Actions tab of the METplus GitHub webpage.

This workflow is triggered when a release is published on GitHub.
It uses cURL to trigger a Slack message on the DTC-METplus announcements
channel that lists information about the release. A Slack bot was created
through the Slack API and the webhook that generated for the Slack channel
was saved as a GitHub Secret.

.. _cg-ci-testing-workflow:

Testing Workflow
================

Name
----

> The name of a workflow can be specified to describe an overview of what is run.

Reviewer: More direct.
```suggestion
The workflow name provides a brief description of what is run.
```

The following line in the testing.yml file::

    name: Testing

defines the workflow identifier that can be seen from the Actions tab on the
METplus GitHub page.

.. figure:: figure/gha-workflow-name.png

Event Control
-------------

> The **on** keyword is used to determine which events will trigger the workflow

Reviewer: More direct.
```suggestion
The **on** keyword defines which events trigger the workflow
```

to run. There are currently 3 types of events that trigger this workflow:
push, pull_request, and workflow_dispatch.
The jobs that are run in this workflow depend on which event has triggered it.

> There are a lot of jobs that are common to multiple events.

Reviewer: ```suggestion
Many jobs are common to multiple events.
```

Push
^^^^

::

    on:

      push:
        branches:
          - develop
          - develop-ref
          - 'feature_*'
          - 'main_*'
          - 'bugfix_*'
        paths-ignore:
          - 'docs/**'

This configuration tells GitHub Actions to trigger the workflow when changes
are pushed to the repository and the following criteria are met:

* The branch is named **develop** or **develop-ref**
* The branch starts with **feature\_**, **main\_**, or **bugfix\_**
* Changes were made to at least one file that is not in the **docs** directory.

Pull Request
^^^^^^^^^^^^

::

      pull_request:
        types: [opened, reopened, synchronize]
        paths-ignore:
          - 'docs/**'

This configuration tells GitHub Actions to trigger the workflow for
pull requests in the repository and the following criteria are met:

* The pull request was opened, reopened, or synchronized.
* Changes were made to at least one file that is not in the **docs** directory.

The **synchronize** type triggers a workflow for every push to a branch
that is included in an open pull request.
If changes were requested in the pull request review,
a new workflow will be triggered for each push.

> To prevent many workflows from being triggered, the pull request

Reviewer: Consider editing with:

To prevent many workflows from being triggered, developers are encouraged to limit the number of pushes for open pull requests. Note that pull requests can be closed until the necessary changes are completed, or [Commit Message Keywords](https://metplus.readthedocs.io/en/feature_675_ci_doc/Contributors_Guide/continuous_integration.html#cg-ci-commit-message-keywords) can be used to suppress the testing workflow.

* **ci-skip-all**: Don't run anything - skip all automation jobs
* **ci-skip-use-cases**: Don't run any use cases
* **ci-skip-unit-tests**: Don't run the Pytest unit tests
* **ci-run-all-cases**: Run all use cases
* **ci-run-diff**: Obtain truth data and run diffing logic for
  use cases that are marked to run
* **ci-run-all-diff**: Obtain truth data and run diffing logic for
  all use cases

.. _cg-ci-get-image:

Create/Update METplus Docker Image
----------------------------------

This job calls the **docker_setup.sh** script (found in .github/jobs).
This script builds a METplus Docker image and pushes it to DockerHub.
The image is pulled instead of built in each test job to save execution time.
The script attempts to pull the appropriate Docker image from DockerHub
(dtcenter/metplus-dev:**<BRANCH_NAME>**) if it already exists so that unchanged

> components of the Docker image do not need to be rebuilt.

Reviewer: Clarify a bit more. Are you saying, for example, that it first pull dtcenter/metplus-dev:develop to get many of the common layers, prior to recreating it from scratch? Is that correct? So the partial rebuild is faster than a complete rebuild?

Author: Yes, that is correct. However, the first step is to pull the MET Docker image, which is a likely piece that would change. If it does, the rest of the steps need to be rerun. If the MET Docker image has not changed in between runs, then I think this logic speeds things up quite a bit.

DockerHub Credentials
^^^^^^^^^^^^^^^^^^^^^

The credentials needed to push images to DockerHub are stored in Secret

> Environment Variables for the repository. These variables are passed

Reviewer: Are these repo-specific secrets or are they now organization-wide secrets? Maybe this detail doesn't matter.

job, then data_assimilation 3 in another job.

Run Use Cases
^^^^^^^^^^^^^

The **use_case_tests** job is duplicated for each use case group using the
strategy -> matrix syntax::

    strategy:
        fail-fast: false
        matrix: ${{fromJson(needs.job_control.outputs.matrix)}}

**fail-fast** is set to false so that the rest of the use case test jobs will
run even when one of them fails. The **matrix** value is a list of use
case categories and indices that is created in the :ref:`cg-ci-job-control`
job. Each value in the list is referenced in the job steps with
**${{ matrix.categories }}**::

    - name: Run Use Cases
      uses: ./.github/actions/run_tests
      id: run_tests
      with:
        categories: ${{ matrix.categories }}

The logic that runs the use cases is contained in a custom GitHub Action
that is found in the METplus repository.

Obtaining Input Data
""""""""""""""""""""

Each use case category has a corresponding Docker data volume that contains
the input data needed to run all of the use cases. The data volume is obtained
from DockerHub and mounted into the container that will run the use cases

> using the **--volumes-from** argument to the **docker run** command.

Reviewer: The --volumes-from is NOT being rendered correctly via RTD. See this [section](https://metplus.readthedocs.io/en/feature_675_ci_doc/Contributors_Guide/continuous_integration.html#obtaining-input-data).
